# Hydrothermal_plume_omics_Zhou_et_al._2020

This GitHub repository contains bioinformatic analyzing methods within the manuscript:

 XXX

The bioRxiv preprint of this manuscript can be found [here](https://www.biorxiv.org/content/10.1101/2020.08.24.253096v1). 

10/6/2020  
Zhichao Zhou   
Karthik Anantharaman  
University of Wisconsin-Madison 

[Anantharaman Lab](https://anantharamanlab.com/)


## Table of Contents:


2. [Phage analysis](#phage analysis)
3. [Metagenome and metatranscriptome mapping](#mapping)
4. [Comparison among three ecosystems](#comparison)
4. [Core microbiome analysis input fasta files](#core_microbiome)
5. [Contact](#contact)


## Explanations


### Phage analysis <a name="phage analysis"></a>
_Dependencies_: Perl v5.22.1, MMseqs2 Version: 7.4e23d, Python 3.5.2

(Higher version of each software should be OK, normally. Detailed information for backward compatibility could be found in the official website for each software )

_Explanation to each step in the pipeline:_  

**Step 1** 01.Get_statistics_from_VIBRANT_result.pl

Get the statistics of phage coding density from VIBRANT result. This will give a table with head line containing: "Plume ID", "Phage sequence number", "Metagenome sequence number", "Phage sequence density (n / 1,000 metagenome sequences)", "Phage nucleotide number", "Metagenome nucleotide number", "Phage nucleotide density (n / 1,000,000 metagenome nucleotides)". Resulted table indicates two phage coding densities (phage sequence density and phage nucleotide density) for each metagenome.

**Step 2** 02.Run_phage_taxonomy.pl and phage_taxonomy_tool_v6.py

Get phage taxonomy at subfamily level. Resulted folder contains two files: "~.VIVID.protein-taxonomy.tsv" and “~.VIVID.virus-taxonomy.tsv”, showing phage protein and phage genome taxonomy results respectively.

**Step 3** 03.Change_seq_head.pl

Change the head of sequences in a given fasta/faa file to 1) replace space to underscore, and cut head by its first "\t", and 2) add plume ID to sequence head. This helps to separate each metagenome, and make tracing each metagenome and each scaffold easier in the downstream analysis.

**Step 4** 04.Dereplicate_phage_proteins.sh

Dereplicate phage proteins first according to their sequence similarity.  The setting was "--min-seq-id 0.95 -c 0.8 -s 7.5 -e 0.001 ", which means sequence identity cutoff set as 95%, coverage cutoff set as 80%, sensitivity   coefficient set as 7.5 (usually just stay at the max of 7.5), and e-value set as 0.001.

**Step 5** 05.Run_mmseqs_to_all_phage_proteins.sh

Run MMseqs2 to cluster all phage proteins. The setting was "--min-seq-id 0.25 -c 0.5 -s 7.5 -e 0.001".

**Step 6** 06.Filter_small_clusters.sh

Filter small clusters which contain only proteins from 10 genomes. This will result a folder named "\*mmseqs_clusters", within which all cluster proteins were provided as "cluster_\*.faa".

**Step 7** 07.Make_genome_list.pl

Firstly, use "ls" to make a list for all cluster protein files in folder "\*mmseqs_clusters" and name the list as "cluster_list.txt". Second, use this script to make "genomes_list.txt" in folder "\*mmseqs_clusters".

**Step 8** 08.Count_cluster_hits_by_genome.sh and count_cluster_hits_by_genome.py

Count cluster hits by genome, and write result into a table with in folder "\*mmseqs_clusters".

**Step 9** 09.Make_metadata.pl

Make metadata table for all phage genomes. It contains columns "Genome", "Taxonomy", "Environment",	"Lysogenic property", "Clusters".

**Step 10** heatmap_plot.R

The R script needs open two input files and produce the heatmap to cluster phage genomes.  The "Count_cluster_hits_by_genome_result.tsv" generated by Step 8 and "Gene_and_taxonomy_metadata.tsv" generated by Step 9 will be used as inputs. The heatmap contains strips of metadata input, dendrogram of genome clusters, and dendrogram of protein clusters.

**Step 11** 10.Get_unknown_phage_proteins_from_each_metagenome.pl

Since the result of clustering all phage genomes are not good (~10k genomes), we try to only analyze the unknown phages (~5k genomes). This script helps to get unknown phage proteins from each metagenome and put them into a new folder.

**Step 12** 11.Dereplicate_phage_proteins.batch_for_each_metagenome.pl

Dereplicate the unknown phage proteins for each metagenome. Settings are the same with those in Step 4.

**Step 13** 12.Run_mmseqs_to_all_phage_proteins.batch_for_each_metagenome.pl

Run MMseqs2 to cluster all unknown phage proteins for each metagenome.  Settings are the same with those in Step 5.

**Step 14** 13.Summarize_protein_diversity.pl

Summarize the unknown phage protein diversity information. Produce a table for each metagenome containing the information of "Unknown phage protein num", "dRep95 unknown phage protein num", "Unknown phage protein cluster num".

The result suggests unknown phages share limited sequence similarity and can hardly be divided into certain groups by clustering phage proteins. Here, we only give the summary information to show the diverse phage protein pattern of unknown phages.



### Metagenome and metatranscriptome mapping <a name="mapping"></a>

Mapping metagenomic and metatranscriptomic reads separately for each environment

_Dependencies_: Perl v5.22.1, Bowtie 2 v2.3.4.1, jgi_summarize_bam_contig_depths (within metaWrap v1.0.2), (pileup.sh) BBmap

(Higher version of each software should be OK, normally. Detailed information for backward compatibility could be found in the official website for each software )

_Explanation to each step in the pipeline:_   

**Step 1**  01.Transcriptom_mapping.Sep_Mapping.in_33.pl

Concatenate all the genes from genomes within each environment to make the mapping reference file. Use the QC-passed and rRNA-filtered metatranscriptomic reads to map against the reference file by Bowtie 2. The template input files, including "Transcriptom_map.mdf.txt" and "MAG_info.txt", are provided in the folder. In "Transcriptom_map.mdf.txt", metatranscriptomic reads are the reads that are labeled with "cDNA" in the second column.

**Step 2**  02.metagenome_mapping.Sep_Mapping.in_33.pl

Concatenate all the genomes within each environment to make the mapping reference file. Use the QC-passed metagenomic reads to map against the reference file by Bowtie 2. The template input files, including "Transcriptom_map.mdf.txt" and "MAG_info.txt", are provided in the folder.  In "Transcriptom_map.mdf.txt", metagenomic reads are the reads that are labeled with "DNA" in the second column.

**Step 3**  03.calculate_metagenome_to_MAG_depth.Sep_Mapping.pl

Calculate depth files based on "sorted.bam" files resulted from Step 2 (for metagenomes). "jgi_summarize_bam_contig_depths"  within metaWrap is used to do the calculation.

**Step 4**  04.pileup_to_calculate_gene_reads_abundance_for_MetaT.Sep_Mapping.sh

Calculate depth files based on "sorted.bam" files resulted from Step 1 (for metatranscriptomes). "pileup.sh" within BBmap is used to do the calculation.

**Step 5**  05.calculate_MAG_average_coverage.Sep_Mapping.pl

Calculate each MAG average coverage based on the result from Step 3. Resulted files are named as "*.MAG_average_coverage.txt". We normalize each metagenomic datasets to the size of 100M reads.

**Step 6**  06.parse_pileup_info.Sep_Mapping.pl

Parse the pileup outputs from Step 4.

**Step 7**  07.calculate_MetaT_RPKM.Sep_Mapping.pl

Calculate each gene average MetaT expression level based on the result from Step 6. Resulted files are named as "*.MetaT.RPKM.txt", and all the resulted values are in RPKM.

### Comparison among three ecosystems <a name="comparison"></a>

_Dependencies_: Perl v5.22.1, R version 4.0.2, R Studio Version 1.3.1073

(Higher version of each software should be OK, normally. Detailed information for backward compatibility could be found in the official website for each software )

_Explanation to each step in the pipeline:_    
**Step 1**  01.get_the_statistics_of_group_MetaG_abundance.pl

This script will produce a table named "MetaG.10times.3_ecosystem.Group2MetaG_id_MetaG.xls" which contains the total gene coverage (10 times of the original gene coverage) of each microbial group. The gene coverage was used by multiplying 10, because the original numbers are too small. After this, we make a new table, named "MetaG.10times.3_ecosystem.Group2MetaG_id_MetaG.integer.txt" which transfers all the numbers from the previous table into integers. This will be used as the input for R script "DESeq2.MAG.coverage.3ecosystem.R ". The template input files, including "MAG_average_coverage.10times.txt" and "MAG_info.txt", are provided in the folder.

**Step 2**  02.get_the_statistics_of_group_MetaG_abundance_for_each_environment.pl

This script will produce two tables named "MAG_average_coverage.Group2Row_Mean_MetaG.xls" and "MAG_average_coverage.Group2Row_Mean_MetaG_each_bin.xls". This script is the same as the Step 1 script, while it is adjusted for each environment.  The template input files, including "MAG_average_coverage.row_mean.simple.txt" and "MAG_info.txt", are provided in the folder for each environment. "MAG_average_coverage.row_mean.simple.txt" was made by taking the mean value of each row (including the background and plume mean values) for each environment based on "MAG_average_coverage.10times.txt". 

**Step 3**  03.parse_fun_normalized_abundance.Sep_Mapping.MetaT.v2.pl

Calculate normalized coverage for each function trait. The input each genome metagenome coverage files are from Step 5 of Metagenome and metatranscriptome mapping (values are transformed into percentage for each environment). Resulted files are names as "*.Fun2MetaG_abundance.txt". All the results are then combined into "Functional_analysis_summary.mdf.txt" (numbers are transferred into integers), which is provided in the folder.

**Step 4**  04.get_taxa_for_enriched_functions.pl

Get the microbial community contribution information to the enriched functions in each environment. The "enriched functions" (refer to "Enriched_functions.txt") are functions that are significantly enriched in each environment calculated by Step 3 and Rscript 2. The template input files, including "MAG_info.txt", "Fun_result.txt", "MetaG.10times_coverage.txt", and "Enriched_functions.txt", are provided in the folder. Resulted files are named as "Enriched_fun_micro_grp.*.txt".

**Step 5**  05.get_taxa_for_major_functions.pl

Get the microbial community contribution information to the major functions in each environment. The major functions include functions that are in the categories of carbon fixation, denitrification, sulfur cycling, hydrogen oxidation, and methane oxidation. The template input files, including "MAG_info.txt", "Fun_result.txt", "MetaG.10times_coverage.txt", and "Major_functions.txt", are provided in the folder. Resulted files are named as "Major_fun_micro_grp.*.txt".



*Rscripts*：

**Rscript 1** DESeq2.MAG.coverage.3ecosystem.R 
Perform DESeq2 analyses to compare the microbial community difference among different background and plume environment pairs, including B.Cym vs B.Lau, P.Cym vs P.Lau, P.Cym vs P.GyBn, and P.GyBn vs P.Lau. The Log2 Fold Change and *P*-value are calculated for each microbial group and included in the result.

**Rscript 2** DESeq2.MAG.function_analysis.3ecosystem.R 

Perform DESeq2 analyses to compare the microbial function difference among different background and plume environment pairs, including B.Cym vs B.Lau, P.Cym vs P.Lau, P.Cym vs P.GyBn, and P.GyBn vs P.Lau. The Log2 Fold Change and *P*-value are calculated for each microbial group and included in the result.

**Rscript 3** make_percentage_table.R

Make percentage table based on input table



### Core microbiome analysis input fasta files <a name="core_microbiome"></a>

We put all the fasta files into the folder of "Core_microbiome_analysis_input_fasta_files".  One could recover and extract the tar.gz file like this:

cat fasta_files.tar.gz* > fasta_files.tar.gz ; tar xzf fasta_files.tar.gz

The routine [QIIME](http://qiime.org/) 16S rRNA gene analyzing method was used to interpret the microbial community pattern.  



​    

### Contact <a name="contact"></a>

Zhichao Zhou,  zczhou2017@gmail.com & zzhou388@wisc.edu

Karthik Anantharaman,   karthik@bact.wisc.edu 